### ISSUE TITLE
 How are the .txt files in trajnetbaselines/center_lines produced?

### ISSUE THREAD

**Issue by ningwak** [vita-epfl/RRB#3]
I didn't find them in the downloaded Interaction Dataset.

**Comment by MohammadHossein-Bahari** [comment]
Hello,
First, note that the Interaction dataset provides lanes' information in "osm" format, but we could not leverage them. The reason was that they provide lanes of the road, while we require centerlines, the center of the road that vehicles usually follow (some other datasets like Argoverse include centerline information). Anyway, as the conversion from what they provide to centerline is an error-prone engineering task (may not seems so, but is not easy at all :) ), we extracted these .txt files ourselves, by running clustering algorithms on the trajectory data and then averaging each cluster. 
Let me know if you have further questions.

**Comment by ningwak** [comment]
Thank you very much! Is the clustering converter included in the repository? In fact I am trying to apply the model learned from Interaction Dataset to another simulation environment so I need to understand how these .txt files are structured to poduce them acorrding to the road layout of that environment. However, when I downloaded and opened the .txt files I found them garbled while it didn't raise error when I ran the training and evaluation.

**Comment by MohammadHossein-Bahari** [comment]
No, it's not included.
The files are in pickle format. To load it, use the following code (written [here](https://github.com/vita-epfl/RRB/blob/e77168dd7d199b7d5396da21a1f1011b6350ed0e/trajnetbaselines/trajnetbaselines/trainer.py#L56)) :

```
with open("file_address" + ".txt" , "rb") as fp:
     array = pickle.load(fp) 
```

The format is [num_lanes, 2, num_points] where 2 is for x,y. So, if a scene has 4 lanes and the maximum number of points for lanes is 1000, the dimension is [4,2,1000].

**Comment by ningwak** [comment]
I  tried to print the loaded file, but it seem to have more dimensions than you said. I am still a bit confused about what each dimension means.[
<img width="462" alt="微信图片_20210623102341" src="https://user-images.githubusercontent.com/41559363/123063046-58897980-d40d-11eb-8e48-181f817fe79a.png">
](url)

**Comment by ningwak** [comment]
And how are the segmented images produced?

**Comment by MohammadHossein-Bahari** [comment]
The format is the same as what I mentioned before. For example, the file you showed here, "DR_USA_Intersection_GL.txt", has an array of [38, 2, 1806]. So it includes 38 lanes, and each lane has 1806 points.

Regarding your second question, the segmented images are produced using [labelbox ](https://labelbox.com/) tool manually.

**Comment by ningwak** [comment]
Now I am trying to define my own scenerio and vehicle paths. I noticed inside "segmentedImgs" folder there is a file named offset.txt and I am wondering what "offset" means here and how to define the two offset values for a scenerio.

**Comment by MohammadHossein-Bahari** [comment]
Sorry for my late reply. I'm glad that you are progressing on your own data!
The "offset" and "scale" values are there to convert vehicle positions (which are x,y global coordinates) to the pixel domain of the scene image. I required this conversion to be able to normalize scene images according to the location of the main agent. Then, I convert values back to the global coordinate.

**Comment by ningwak** [comment]
Thank you very much for your reply! So are the (x, y) positions printed above after being loaded from the .txt files in global coordinates or the pixel domain?

**Comment by MohammadHossein-Bahari** [comment]
Following my previous answer, the conversion from the global domain to the pixel domain is done [here](https://github.com/vita-epfl/RRB/blob/e77168dd7d199b7d5396da21a1f1011b6350ed0e/trajnetbaselines/trajnetbaselines/trainer.py#L277).
Regarding your question, the centerlines in the .txt files are in the pixel domain.

**Comment by ningwak** [comment]
After the conversion are the positions converted to the pixel domain in normalized coordinates? If so, how can they correspond to the center lines which from my understanding are in pixel domain and global coordinates?

**Comment by lxpr** [comment]
Now I am testing using hard coded data. The center line is set as 
`torch.tensor([[[800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 800, 10000],
                            [1, 100, 200, 300, 400, 500, 600, 700, 800, 900 ,1000, 1100, 1200, 1300, 1400, 1500, 1600, 1700, 1800, 1900, 10000]]])`
And the .ndjson file is 
`{"scene": {"id": 35089, "p": 10239, "s": 44875, "e": 44945}}
{"track": {"f": 44875, "p": 10239, "x": 800, "y": 90}}
{"track": {"f": 44880, "p": 10239, "x": 800, "y": 95}}
{"track": {"f": 44885, "p": 10239, "x": 800, "y": 100}}
{"track": {"f": 44890, "p": 10239, "x": 800, "y": 108}}
{"track": {"f": 44895, "p": 10239, "x": 800, "y": 118}}
{"track": {"f": 44900, "p": 10239, "x": 800, "y": 125}}
{"track": {"f": 44905, "p": 10239, "x": 800, "y": 130}}
{"track": {"f": 44910, "p": 10239, "x": 800, "y": 135}}
{"track": {"f": 44915, "p": 10239, "x": 800, "y": 140}}
{"track": {"f": 44920, "p": 10239, "x": 800, "y": 145}}
{"track": {"f": 44925, "p": 10239, "x": 800, "y": 150}}
{"track": {"f": 44930, "p": 10239, "x": 800, "y": 155}}
{"track": {"f": 44935, "p": 10239, "x": 800, "y": 160}}
{"track": {"f": 44940, "p": 10239, "x": 800, "y": 165}}
{"track": {"f": 44945, "p": 10239, "x": 800, "y": 170}}`
with only one scene and one vehicle. I set scale as 8 and offset as (87.5, 700) to make the positions after convertion on the center line. I am producing strange predictions like 
`[[Row(frame=44900, pedestrian=10239, x=tensor(807.5000), y=tensor(19.2500)), Row(frame=44905, pedestrian=10239, x=tensor(806.8750), y=tensor(1
9.2500)), Row(frame=44910, pedestrian=10239, x=tensor(806.6250), y=tensor(19.2500)), Row(frame=44915, pedestrian=10239, x=tensor(806.5000), y=tensor(19.250
0)), Row(frame=44920, pedestrian=10239, x=tensor(806.5000), y=tensor(19.2500)), Row(frame=44925, pedestrian=10239, x=tensor(806.5000), y=tensor(19.2500)),
Row(frame=44930, pedestrian=10239, x=tensor(806.7500), y=tensor(19.2500)), Row(frame=44935, pedestrian=10239, x=tensor(807.6250), y=tensor(19.2500)), Row(f
rame=44940, pedestrian=10239, x=tensor(807.7500), y=tensor(19.2500)), Row(frame=44945, pedestrian=10239, x=tensor(807.7500), y=tensor(19.2500))]]`

**Comment by MohammadHossein-Bahari** [comment]
After conversion, the positions are in the pixel domain. Then, we normalize them based on the position and direction of ego-vehicle [here](https://github.com/vita-epfl/RRB/blob/e77168dd7d199b7d5396da21a1f1011b6350ed0e/trajnetbaselines/trajnetbaselines/trainer.py#L280). The centerlines are normalized in the next line of code [here](https://github.com/vita-epfl/RRB/blob/e77168dd7d199b7d5396da21a1f1011b6350ed0e/trajnetbaselines/trajnetbaselines/trainer.py#L282). So, we have first the positions and centerlines in the pixel domain, then normalize them.

**Comment by swb19** [comment]
> Thank you very much! Is the clustering converter included in the repository? In fact I am trying to apply the model learned from Interaction Dataset to another simulation environment so I need to understand how these .txt files are structured to poduce them acorrding to the road layout of that environment. However, when I downloaded and opened the .txt files I found them garbled while it didn't raise error when I ran the training and evaluation.

hello, could you please tell me where I downloaded these .txt files?

**Comment by ningwak** [comment]
> > Thank you very much! Is the clustering converter included in the repository? In fact I am trying to apply the model learned from Interaction Dataset to another simulation environment so I need to understand how these .txt files are structured to poduce them acorrding to the road layout of that environment. However, when I downloaded and opened the .txt files I found them garbled while it didn't raise error when I ran the training and evaluation.
> 
> hello, could you please tell me where I downloaded these .txt files?

They are inside this folder. https://github.com/vita-epfl/RRB/tree/main/trajnetbaselines/center_lines

**Comment by swb19** [comment]
> > > Thank you very much! Is the clustering converter included in the repository? In fact I am trying to apply the model learned from Interaction Dataset to another simulation environment so I need to understand how these .txt files are structured to poduce them acorrding to the road layout of that environment. However, when I downloaded and opened the .txt files I found them garbled while it didn't raise error when I ran the training and evaluation.
> > 
> > 
> > hello, could you please tell me where I downloaded these .txt files?
> 
> They are inside this folder. https://github.com/vita-epfl/RRB/tree/main/trajnetbaselines/center_lines

Thank you very much! I have found them! But I have another question, in the trajnetdataset/trajnetdataset/convert.py, some other.txt files are required for train_data, such as 'DJI_0001-Roseville-annot1Point.txt', do you know how to find them?

**Comment by MohammadHossein-Bahari** [comment]
> > > > Thank you very much! Is the clustering converter included in the repository? In fact I am trying to apply the model learned from Interaction Dataset to another simulation environment so I need to understand how these .txt files are structured to poduce them acorrding to the road layout of that environment. However, when I downloaded and opened the .txt files I found them garbled while it didn't raise error when I ran the training and evaluation.
> > > 
> > > 
> > > hello, could you please tell me where I downloaded these .txt files?
> > 
> > 
> > They are inside this folder. https://github.com/vita-epfl/RRB/tree/main/trajnetbaselines/center_lines
> 
> Thank you very much! I have found them! But I have another question, in the trajnetdataset/trajnetdataset/convert.py, some other.txt files are required for train_data, such as 'DJI_0001-Roseville-annot1Point.txt', do you know how to find them?

The convert.py file processes the annotation txt files provided by the dataset. I have uploaded the processed files in RRB/trajnetdataset/output_interaction_* folders. You do not need to run the convert.py file.
